### ISSUE TITLE
[Windows] Unable to uninstall Elastic Agent after it upgrades and the watcher is still running

### ISSUE THREAD

**Issue by blakerouse** [elastic/elastic-agent#3039]
It is not possible to get Elastic Agent to uninstall on Windows after the Elastic Agent has just upgraded. This is because the watcher is still running and its writing its log files to the directory that the Elastic Agent is trying to remove.

I believe that on all platforms the Elastic Agent should kill the watcher if its performing an uninstall, as the watcher is no longer required as the Elastic Agent is getting removed from the host.

```
 fixture.go:345: >> running agent with: [C:\Program Files\Elastic\Agent\elastic-agent.exe uninstall --force]
    fixture_install.go:114: 
        	Error Trace:	C:/Users/windows/agent/pkg/testing/fixture_install.go:114
        	            				C:/Program Files/Go/src/testing/testing.go:1150
        	            				C:/Program Files/Go/src/testing/testing.go:1328
        	            				C:/Program Files/Go/src/testing/testing.go:1570
        	Error:      	Received unexpected error:
        	            	error running uninstall command: exit status 1
        	Test:       	TestFleetManagedUpgrade
        	Messages:   	uninstalling agent failed. Output: "Error: failed to remove installation directory (C:\\Program Files\\Elastic\\Agent): remove C:\\Program Files\\Elastic\\Agent\\data\\elastic-agent-4ac18b\\logs\\elastic-agent-watcher-20230707-1.ndjson: The process cannot access the file because it is being used by another process.\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.10/fleet-troubleshooting.html\n"
--- FAIL: TestFleetManagedUpgrade (231.06s)
```

**Comment by cmacknz** [comment]
This is available starting in 8.11

**Comment by cmacknz** [comment]
The actual fix commit was https://github.com/elastic/elastic-agent/commit/7e86d24cd336145b9641f621a3de10ed3af0c9fe

**Comment by maggieghamry** [comment]
Thank you @cmacknz !

**Comment by maggieghamry** [comment]
@cmacknz the customer

**Comment by maggieghamry** [comment]
@cmacknz the customer is indicating that this issue is still occurring in `8.11.0` on ESS, and they are working to escalate this issue, indicating there may be another bug, or this is persistiing. 

Can you please let me know if there may be another bug related? 

> 8.11.1 was released, but still cannot upgrade from the previous release (in this case 8.11.0) to 8.11.1. Prior versions you can type 8.11.1 in in the Upgrade window, but no option for the previous release.
> 
> As for the actual reason for creating this ticket, the agent uninstall still fails with 8.11.1 on the last step (can't remove folder because it is in use). I have waited 10 minutes per Release Notes between the Upgrade and attempting Uninstall.
> 
> The issue is not fixed with 8.11.0, or there is another similar bug:
> 
> 
> PS [C:\Program](file:///C:/Program) Files\Elastic\Agent> .\elastic-agent.exe uninstall -f
> Stopping service... DONE
> Stopping upgrade watcher (3860)........ DONE
> Removing service... DONE
> Removing install directory.......................... FAILED
> Error: failed to remove installation directory ([C:\Program](file:///C:/Program) Files\Elastic\Agent): timed out while removing "[C:\\Program](file:///C://Program) Files[\\Elastic\\Agent](file://elastic//Agent)". Last error: remove [C:\Program](file:///C:/Program) Files\Elastic\Agent: The process cannot access the file because it is being used by another process.
> For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.11/fleet-troubleshooting.html
> PS [C:\Program](file:///C:/Program) Files\Elastic\Agent>
> 
> When I look, the service is gone and the [C:\Program](file:///C:/Program) Files\Elastic\Agent directory is empty.

> It almost looks like it needs a wait in there somewhere.

They have also run into when attempting to upgrade to `8.11.1` and cannot type it in (they are in ESS - deployment https://admin.found.no/deployments/2ca7143295ce44fcb54f866e1f5f1c8e)
<img width="714" alt="image" src="https://github.com/elastic/elastic-agent/assets/46542915/aa215950-518f-4d9c-aa90-adaf2c260afd">

**Comment by cmacknz** [comment]
Closing again, believe the reason for reopening was a separate problem.

**Comment by FlorianHeigl** [comment]
Just had a few of these on 8.11.1
Normally you can delete the file seconds after this is completed. Certainly some race condition?

**Comment by cmacknz** [comment]
Yes there are a few places where we were incorrectly logging to files in the directory we were also attempting to remove. On Windows you get an error like this, the specific file varies a bit.

**Comment by ankitsheoran1** [comment]
Hey i was also testing uninstall with elastic 8.12.2 on windows and faced same issues

**Comment by pierrehilbert** [comment]
@krishhteja and @ankitsheoran1 Thanks for raising your issue.
Could you please give us more details about your setup to help us to reproduce the issue?

**Comment by ankitsheoran1** [comment]
> @krishhteja and @ankitsheoran1 Thanks for raising your issue. Could you please give us more details about your setup to help us to reproduce the issue?

yeah sure we are using elastic version 8.12.0 . We are   installing agent on window 10 in steps 
```
1. wget https://artifacts.elastic.co/downloads/beats/elastic-agent/elastic-agent-8.12.0-windows-x86_64.zip 
2. elastic-agent.exe install -f --url=$FleetURL --enrollment-token=$EnrollmentToken
```
With help of above steps we able to successfully install elastic agent . 

Then we try to uninstall by command 
`1. elastic-agent uninstall -f . `
This successfully uninstalled but at end it throws error , I am adding a screenshot here 
![elastic-agent](https://github.com/elastic/elastic-agent/assets/33404702/4b7d9e52-85eb-4892-9a2d-a15bd614cdf8). 
We also trying first to stop elastic-agent and then uninstall but nothing works

**Comment by cmacknz** [comment]
That error should be fixed in 8.13.0 with https://github.com/elastic/elastic-agent/pull/4108

We found that our uninstall command was sometimes attempting to log to disk as it was trying to delete the agent directory.

We were able to regularly reproduce these errors in our integration tests and haven't seen them since that fix was made.
